### PR TITLE
Fix slicer upload in some cases

### PR DESCRIPTION
**Pull request by smallhive**
The case when the reader returns io.EOF with the last data chunk. Slicer was closing the writer without payload writing

### CODE REVIEW

**Review by carpawell — COMMENTED** [review]
lol. Does it mean that RC10 isn't perfect again in term of `Slicer`? How do you found it?

Commit's body does not end with a period.

**Review by cthulhu-rider — APPROVED** [review]

**Comment by smallhive** [comment]
> lol. Does it mean that RC10 isn't perfect again in term of `Slicer`? How do you found it?
> 
> Commit's body does not end with a period.

Period is fixed. I'm updating our gates. Inside HTTP one I updated all upload places to slicer using. 
One test started to fail: debug, investigation, found, fixed, profit :smiley:

**Review by roman-khimov — CHANGES_REQUESTED** [review]
It's disappointing to see this given that exactly this issue had been discussed numerous times in slicer-related PRs. Yet at the same all PRs were reviewed successfully before the respective merges.

Let's hope the new test won't let it happen again.

**Review by roman-khimov — APPROVED** [review]

File: object/slicer/slicer_test.go [hunk 1]
```diff
@@ -269,6 +269,23 @@ func testSlicer(t *testing.T, size, sizeLimit uint64) {
 	}
 }
 
+// eofOnLastChunkReader is a special reader for tests. It returns io.EOF with the last data portion.
+type eofOnLastChunkReader struct {
+	Payload []byte
+	point   int
+}
+
+func (l *eofOnLastChunkReader) Read(p []byte) (int, error) {
```
Comment (cthulhu-rider) [inline object/slicer/slicer_test.go:278]: lets also return `0, nil` to test all `io.Reader` doc statements? For example 
```go
func (l *eofOnLastChunkReader) Read(p []byte) (int, error) {
  if !l.readEmpty {
    l.readEmpty = true
    return 0, nil
  }
```

Comment (smallhive) [reply object/slicer/slicer_test.go:279]: Extended with new test

Comment (cthulhu-rider) [reply object/slicer/slicer_test.go:279]: does it ever return `0, nil`?

Comment (smallhive) [reply object/slicer/slicer_test.go:279]: No, depending on the condition

Comment (cthulhu-rider) [reply object/slicer/slicer_test.go:279]: lets add then, it's simple (see initial comment) and tests possible `io.Reader` behavior

Comment (smallhive) [reply object/slicer/slicer_test.go:282]: Updated

**Pull request merged**